WordPress/gutenberg
# Add CodeMirror to the HTML block

### 👀 What this is

Adds CodeMirror to the HTML block. This means that a user can type their custom HTML using a fully featured text editor with linting and syntax highlighting:

![codemirror](https://user-images.githubusercontent.com/612155/34701754-10693cbe-f53e-11e7-9ca6-5c98bf4d6200.gif)

### 🤔 How this works

[CodeMirror was added to core in WordPress 4.9](https://make.wordpress.org/core/2017/10/22/code-editing-improvements-in-wordpress-4-9/) which means we can enhance a `<textarea>` simply by calling `wp.codeEditor.initialize()`.

To make CodeMirror easier to use within React, I introduced a `<CodeEditor>` component which acts as a drop-in replacement for a `<textarea>`.

I've also incorporated some of @westonruter's work in https://github.com/WordPress/gutenberg/pull/1625 by using `SandBox` within our HTML block. This makes previewing a HTML block more robust. Specifically, you can now insert HTML that blocks rendering, e.g.:

```html
<script>document.write( 'Hey!' );</script>
```

To avoiding adding a whopping 1.9 MB to our assets, I've implemented a very simple mechanism to lazily load the assets that are normally queued by calling `wp_enqueue_code_editor()`. 

### ✅ How to test

1. Create a HTML block
2. Preview it
3. Save it, view the page
4. Check that you can preview HTML that blocks rendering, e.g. `<script>document.write( 'Hey!' );</script>`
4. Log in as an author and test that you get lint warnings when using e.g. the `<script>` tag
